filters/auth/grantlogout.go

```diff
@@ -0,0 +1,202 @@
+	accessTokenType           = "access_token"
+	unsupportedTokenTypeError = "unsupported_token_type"
+)
+
+type grantLogoutSpec struct {
+	config OAuthConfig
+}
+
+type grantLogoutFilter struct {
+	config OAuthConfig
+}
+
+type revokeErrorResponse struct {
+	Error            string `json:"error"`
+	ErrorDescription string `json:"error_description"`
+}
+
+type basicAuthCredentials struct {
+	clientId     string
+	clientSecret string
+}
+
+func (*grantLogoutSpec) Name() string { return GrantLogoutName }
+
+func (s *grantLogoutSpec) CreateFilter([]interface{}) (filters.Filter, error) {
+	return &grantLogoutFilter{
+		config: s.config,
+	}, nil
+}
+
+func (f *grantLogoutFilter) getBasicAuthCredentials() (*basicAuthCredentials, error) {
+	clientID := f.config.GetClientID()
+	if clientID == "" {
+		return nil, errors.New("failed to create token revoke auth header: no client ID")
+	}
+
+	clientSecret := f.config.GetClientSecret()
+	if clientSecret == "" {
+		return nil, errors.New("failed to create token revoke auth header: no client secret")
+	}
+
+	return &basicAuthCredentials{
+		clientId:     clientID,
+		clientSecret: clientSecret,
+	}, nil
+}
+
+func getErrorResponse(response *http.Response) *revokeErrorResponse {
+	buf, err := ioutil.ReadAll(response.Body)
+	if err == nil && buf != nil {
+		var errorResponse revokeErrorResponse
+		err = json.Unmarshal(buf, &errorResponse)
+
+		if err == nil {
+			return &errorResponse
+		}
+	}
+
+	return nil
+}
+
+func responseToError(response *http.Response, tokenType string) error {
+	errorResponse := getErrorResponse(response)
+	if errorResponse.Error == unsupportedTokenTypeError && tokenType == accessTokenType {
+		// Provider does not support revoking access tokens, which can happen according to RFC 7009.
+		// In that case this is not really an error.
+		return nil
+	}
+	return fmt.Errorf(
+		"%s revocation failed: %d %s",
+		tokenType,
+		response.StatusCode,
+		errorResponse.Error+": "+errorResponse.ErrorDescription,
+	)
+}
+
+func (f *grantLogoutFilter) revokeTokenType(tokenType string, token string) error {
+	revokeURL, err := url.Parse(f.config.RevokeTokenURL)
+	if err != nil {
+		return err
+	}
+
+	query := revokeURL.Query()
+	for k, v := range f.config.AuthURLParameters {
+		query.Set(k, v)
+	}
+	revokeURL.RawQuery = query.Encode()
+
+	body := url.Values{}
+	body.Add(revokeTokenKey, token)
+	body.Add(revokeTokenTypeKey, tokenType)
+
+	revokeRequest, err := http.NewRequest(
+		"POST",
+		revokeURL.String(),
+		strings.NewReader(body.Encode()))
+
+	if err != nil {
+		return err
+	}
+
+	authCredentials, err := f.getBasicAuthCredentials()
+	if err != nil {
+		return err
+	}
+
+	revokeRequest.SetBasicAuth(authCredentials.clientId, authCredentials.clientSecret)
+	revokeRequest.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+
+	revokeResponse, err := f.config.AuthClient.Do(revokeRequest)
+	if err != nil {
+		return err
+	}
+	defer revokeResponse.Body.Close()
```

Comment:
There is a catch here, closing is not enough, the body should be read-out. I guess there are quite some other places in the codebase but this one we can fix on the spot (read it and pass buffer to responseToError instead of response).

See https://github.com/zalando/skipper/pull/1511 as an example

Reply:
That is good to know! Fixed.